hier.c

```diff
@@ -27,8 +27,11 @@ int mode;
   }
 }
 
-void hier()
+void hier(queue,everythingelse)
+int queue;
+int everythingelse;
```

Comment:
Passing flags as you're doing here can be a so-called 'code smell.'  In this case there is an extra tell: you have a flag for whether or not to install the queue, and a flag for whether or not to not install the queue (With my attendant appreciation in not calling the variable notqmail!).

If one is faced with passing flags to a function it's natural to ask whether the function should refactored instead.  You've added two flags, creating four combinations, one of which makes this function effectively a no-op--such combination reasonably never appearing in the code.  Further, blocks of lines are now if-guarded, which should create a space-only diff in every line--you've taken care here not to do that, but by "squeezing in" an extra level of indent.

Instead of passing flags to hier(), I think you create two functions: hier() and hier_queue().  For hier(1,1) (instcheck, instchown), call both hier() and hier_queue().  For hier(0,1) (instpackage), call hier().  For hier(1,0) (instqueue), call hier_queue().  Pulling hier_queue() out of hier() accomplishes exactly the same thing adding flags does, but in a way that retains the current complexity of this/these subroutines--no extra choices are made at runtime--consistent with these parameters being hard-coded at each calling site.

Reply:
I’d prefer the code in the state you describe, but aimed for an intermediate state that seemed less likely to break patches. I haven’t tested whether that’s true, though.

Reply:
I don't consider the rationale of not breaking patches sufficient justification to put in constant-at-build time flags to a function.  Quite reasonably possible we'll split all of hier() in to the 8 top-level directory-specific functions.  Absolutely impossible this function signature survives the 1.08 series--it is unreasonably overfit for the reasons I describe here.

Reply:
Hmmm. I'm worried we're already slipping down a slope of making common patches non-trivial for users to apply. (For pkgsrc, I've noticed a need to rebase the TLS and SRS patches onto notqmail.) I can't point at data to quantify the barrier that 1.07 will present to users trying to update; I'm just worried about it. But we have ideas for mitigating this, such as the patch branches you've created in the repo. So maybe I need to set my worry aside, watch how it goes for people when they encounter 1.07, and recalibrate accordingly. Let me refactor as you suggest.